### ISSUE TITLE
Stabilize old experimental/preview rules

### ISSUE THREAD

**Issue by inoa-jboliveira** [astral-sh/ruff#25299]
### Summary

I'd like to drop the `preview` option from my pyproject but I do make use of several rules that have been marked as experimental since forever.

The push comes from the format changes that are happening and it will be kinda large reformat in a project with a couple MLOC.

So I am both stuck with a certain version of ruff and stuck with rules that cannot be used.

Yes, I could drop said rules, but several did catch bugs early on that it is more risky to remove them than to never upgrade ruff.

Example: most of the pycodestyle `E` rules are marked `Rule has been in preview since v0.0.269`. 

What I believe is the problem here is that there is no rule progression process. They are born and instantly forgotten. Can this be fixed?

Many thanks

**Comment by ntBre** [comment]
I think we could do a bit better about sharing this information publicly, but as part of each minor release process, we go through each of the preview rules and make a decision whether or not to stabilize the rule. https://github.com/astral-sh/ruff/issues/19487 is one case where I turned the internal findings into an issue, but I haven't done that in general. The typical criteria we consider are:
- has the rule been in preview long enough? (this typically means one minor release cycle of ~90 days)
- have there been any "recent" issues or significant changes to the rule? (again ~90 days)
- does stabilizing the rule produce a "reasonable" ecosystem report? (https://github.com/astral-sh/ruff/pull/16657 is one example of a "noisy" report that made us reconsider)

For pycodestyle in particular, we've been hesitant to stabilize them for the reasons discussed in https://github.com/astral-sh/ruff/issues/23903#issuecomment-4041999990 and further links there. However, I'm actively working on #1774 again, and my current proposal includes a top-level Formatting category that would give us a very natural place for many of these preview rules.

If there are other specific rules you'd like to see stabilized, feel free to open issues to be tracked in the release milestone. 0.16 should be coming up before long.

(Maybe we can reuse this issue for that purpose, especially if there's a long list of individual rules 😄)

**Comment by inoa-jboliveira** [comment]
Hi @ntBre thank you for such swift reply.

For example, this is the list of things that show up if I disable preview and run latest ruff via `uvx ruff check` 

```
warning: Selection `E111` has no effect because preview is not enabled.
warning: Selection `E112` has no effect because preview is not enabled.
warning: Selection `E304` has no effect because preview is not enabled.
warning: Selection `B901` has no effect because preview is not enabled.
warning: Selection `B903` has no effect because preview is not enabled.
warning: Selection `FURB131` has no effect because preview is not enabled.
warning: Selection `FURB145` has no effect because preview is not enabled.
warning: Selection `FURB152` has no effect because preview is not enabled.
warning: Selection `FURB154` has no effect because preview is not enabled.
warning: Selection `FURB164` has no effect because preview is not enabled.
warning: Selection `FURB180` has no effect because preview is not enabled.
warning: Selection `RUF029` has no effect because preview is not enabled.
warning: Selection `RUF038` has no effect because preview is not enabled.
warning: Selection `RUF047` has no effect because preview is not enabled.
warning: Selection `RUF054` has no effect because preview is not enabled.
warning: Selection `RUF055` has no effect because preview is not enabled.
warning: Selection `RUF056` has no effect because preview is not enabled.
warning: Selection `RUF063` has no effect because preview is not enabled.
warning: Selection `RUF066` has no effect because preview is not enabled.
```

An those are only the ones I manually included. There are probably a bunch that are hidden in some range like `TID` where the preview rules would be simply ignored.

I understand some rules are of lesser value than others and not every single one is a must for me, but going through them is time consuming.

Would it be possible if you did like some script that got any preview rule unchanged since like 4-5 releases (e.g. 0.10 or 0.11) and stabilize it, check the noise and target for merging if low? Except the `E` rules as you put there is a different issues and I fully agree they should be moved to a `FMT` section or something.

Every time you make a minor release, this script could run and do provide insight on what could be stabilized next

**Comment by ntBre** [comment]
> Would it be possible if you did like some script that got any preview rule unchanged since like 4-5 releases (e.g. 0.10 or 0.11) and stabilize it, check the noise and target for merging if low?

Yes, this is the process I use for every minor release.

**Comment by ntBre** [comment]
Closing this as resolved. We stabilized some of the rules requested here in 0.16, and I opened #27051 as a tracking issue with a summary of my findings for future releases.
